File: android/src/main/kotlin/com/example/review_manager/ReviewManagerPlugin.kt
package com.example.review_manager

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

private const val ChannelName = "review_manager"

class ReviewManagerPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {

  //region Variables
  private lateinit var channel : MethodChannel
  private var activity: Activity? = null

  //endregion

  //region FlutterPlugin
  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, ChannelName)
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (methodsFrom(call.method)) {
      Method.RequestReviewIfAppropriate -> {
        val activity = this.activity ?: return
        AppReviewHelper.requestReviewIfAppropriate(activity, result);
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
  //endregion

  //region ActivityAware
  private fun disposeActivity() {
    activity = null
  }

  override fun onDetachedFromActivity() {
    disposeActivity()
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    onAttachedToActivity(binding)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    onDetachedFromActivity()
  }
  //endregion
}

File: darwin/Classes/ReviewManagerPlugin.swift
import Flutter
import UIKit

private let ChannelName = "review_manager"

public class ReviewManagerPlugin: NSObject, FlutterPlugin {
  
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: ChannelName, binaryMessenger: registrar.messenger())
    let instance = ReviewManagerPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }
  
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    
    let method = Method(rawValue: call.method)
    switch method {
    case .requestReviewIfAppropriate:
      ReviewManager.requestReviewIfAppropriate()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
